### PR TITLE
Perf/zl/remove unnecessary delete to solr idx when crdt or sc

### PR DESCRIPTION
**Pull request by zeeshanlakhani**
Fixes https://bashoeng.atlassian.net/browse/RIAK-1504. 

Improves performance by avoiding an extra delete operation to remove siblings through Solr on the put path when riak datatypes or strong consistency is used. Initial patch/issue was discovered by @drewkerrigan, https://github.com/basho-labs/yokozuna_perf_patch.

### CODE REVIEW

**Comment by Basho-JIRA** [comment]
https://github.com/basho/yokozuna/pull/452 

**_[posted via JIRA by Zeeshan Lakhani]**_

File: src/yz_kv.erl [hunk 1]
```diff
@@ -460,3 +463,131 @@ is_owner_or_future_owner(P, Node, Ring) ->
...
 is_service_up(Service, Node) ->
     lists:member(Service, riak_core_node_watcher:services(Node)).
+
+%% @private
+%%
+%% @doc Check if object has 2.0 CRDT datatype entry or property for
+%%      strong consistency.
+-spec is_datatype_or_consistent(obj()) -> boolean().
+is_datatype_or_consistent(Obj) ->
+    case riak_kv_crdt:value(Obj) of
```
Comment (seancribbs) [inline src/yz_kv.erl]: Similar to the `is_strongly_consistent/1` below, you could check the bucket properties only to determine if it's a CRDT. Calling `riak_kv_crdt:value/1` will perform a deserialization of the value, which is extra work if you are not intending to use the result.

Comment (zeeshanlakhani) [inline src/yz_kv.erl]: Makes sense. Will do. 

Zeeshan Lakhani
programmer | 
software engineer at @basho | 
org. member/founder of @papers_we_love |
twitter => @zeeshanlakhani

**Comment by zeeshanlakhani** [comment]
@borshop will now fail until https://github.com/basho/riak_kv/pull/1081 is reviewed/merged, but eunit tests pass locally.

File: src/yz_kv.erl [hunk 2]
```diff
@@ -460,3 +463,139 @@ is_owner_or_future_owner(P, Node, Ring) ->
...
+            false
+    end.
+
+
+%% @private
+%%
+%% @doc Set yz_solr:index delete operation(s) on write_reason.
+-spec delete_operation(obj(), put|handoff|anti_entropy, [doc()], bkey(), lp()) ->
+                              []|[{id, _}]|[{siblings, _}].
+delete_operation(Obj, put, Docs, BKey, LP) ->
```
Comment (seancribbs) [inline src/yz_kv.erl:501]: Can you explain why `handoff` or `anti_entropy` should result in cleanups? Shouldn't the same reasons apply in those cases as for `put`?

Comment (zeeshanlakhani) [inline src/yz_kv.erl:501]: In a conversation I had w/ @drewkerrigan, it was to keep the delete operation for  `handoff`/`anti_entropy` as cleanup b/c there was an edge case where data-typed keys could actually have a siblings and "one of the test cases the client had was like deleting without using a context and then updating the datatype object."

Comment (seancribbs) [inline src/yz_kv.erl:501]: Understood. I think we should fix the underlying problem in that case, but that should be ok for now.

Comment (zeeshanlakhani) [inline src/yz_kv.erl:501]: Agreed, as there's another issue altogether.

**Comment by seancribbs** [comment]
Holding off final review on this for benchmarks.

**Comment by zeeshanlakhani** [comment]
@seancribbs, yep, will post results here comparing `develop` to this branch.

**Comment by zeeshanlakhani** [comment]
@seancribbs will reopen w/ a new PR... as the rebase w/ the changes to dev got a bit funky.

**Comment by Basho-JIRA** [comment]
Updated fix version to 2.0.6 since the GH was included in release notes by Zeeshan.   

**_[posted via JIRA by Patricia Brewer]**_
